pkg/controller/postgresql/extension/reconciler.go

```
@@ -0,0 +1,210 @@
+
+	t := resource.NewProviderConfigUsageTracker(mgr.GetClient(), &v1alpha1.ProviderConfigUsage{})
+	r := managed.NewReconciler(mgr,
+		resource.ManagedKind(v1alpha1.ExtensionGroupVersionKind),
+		managed.WithExternalConnecter(&connector{kube: mgr.GetClient(), usage: t, newDB: postgresql.New}),
+		managed.WithLogger(l.WithValues("controller", name)),
+		managed.WithRecorder(event.NewAPIRecorder(mgr.GetEventRecorderFor(name))))
+
+	return ctrl.NewControllerManagedBy(mgr).
+		Named(name).
+		For(&v1alpha1.Extension{}).
+		WithOptions(controller.Options{
+			MaxConcurrentReconciles: maxConcurrency,
+		}).
+		Complete(r)
+}
+
+type connector struct {
+	kube  client.Client
+	usage resource.Tracker
+	newDB func(creds map[string][]byte, database string) xsql.DB
+}
+
+func (c *connector) Connect(ctx context.Context, mg resource.Managed) (managed.ExternalClient, error) {
+	cr, ok := mg.(*v1alpha1.Extension)
+	if !ok {
+		return nil, errors.New(errNotExtension)
+	}
+
+	if err := c.usage.Track(ctx, mg); err != nil {
+		return nil, errors.Wrap(err, errTrackPCUsage)
+	}
+
+	// ProviderConfigReference could theoretically be nil, but in practice the
+	// DefaultProviderConfig initializer will set it before we get here.
+	pc := &v1alpha1.ProviderConfig{}
+	if err := c.kube.Get(ctx, types.NamespacedName{Name: cr.GetProviderConfigReference().Name}, pc); err != nil {
+		return nil, errors.Wrap(err, errGetPC)
+	}
+
+	// We don't need to check the credentials source because we currently only
+	// support one source (PostgreSQLConnectionSecret), which is required and
+	// enforced by the ProviderConfig schema.
+	ref := pc.Spec.Credentials.ConnectionSecretRef
+	if ref == nil {
+		return nil, errors.New(errNoSecretRef)
+	}
+
+	s := &corev1.Secret{}
+	if err := c.kube.Get(ctx, types.NamespacedName{Namespace: ref.Namespace, Name: ref.Name}, s); err != nil {
+		return nil, errors.Wrap(err, errGetSecret)
+	}
+
+	if cr.Spec.ForProvider.Database != nil {
```

@linuxbsdfreak In addition to this, I think we need to return an error from `Connect` if `DatabaseRef` or `DatabaseSelector` are configured but `Database` is `nil`. If we don't do that, there's a chance that we will install an extension into the default DB rather  than the DB we want to, if one of the database references or selectors is not resolved instantly.